lib/identity/hostdata/config_reader.rb

```diff
@@ -0,0 +1,106 @@
+        @app_root = app_root
+        @s3_client = s3_client
+        @logger = logger
+      end
+
+      def read_configuration(rails_env, write_copy_to: nil)
+        if write_copy_to && !File.exists?(write_copy_to)
+          FileUtils.mkdir_p(File.dirname(write_copy_to))
+          File.write(write_copy_to, base_configuration.to_yaml)
+          FileUtils.chmod(0o640, write_copy_to)
+        end
+
+        base_configs = base_configuration.dup
+        base_configs.delete('development')
+        base_configs.delete('production')
+        base_configs.delete('test')
+        base_configs.merge(
+          base_configuration[rails_env],
+        ).transform_keys(&:to_sym)
+      end
+
+      private
+
+      def base_configuration
+        @base_configuration ||= deep_merge(
+          deep_merge(default_configuration, app_override_configuration),
+          role_override_configuration,
+        )
+      end
+
+      def default_configuration
+        YAML.safe_load(File.read(File.join(app_root, 'config', 'application.yml.default')))
+      end
+
+      def app_override_configuration
+        local_config_filepath = File.join(app_root, 'config', 'application.yml')
+        raw_configs = if Identity::Hostdata.in_datacenter?
+                        app_secrets_s3.read_file(app_configuration_s3_path)
+                      elsif File.exists?(local_config_filepath)
+                        File.read(local_config_filepath)
+                      end
+        YAML.safe_load(raw_configs || '{}') || {}
+      end
+
+      def role_override_configuration
+        return {} if role_configuration_filename.nil?
+        local_config_filepath = File.join(app_root, 'config', role_configuration_filename)
+
+        raw_configs = if Identity::Hostdata.in_datacenter?
+                        app_secrets_s3.read_file(role_configuration_s3_path)
+                      elsif File.exists?(local_config_filepath)
+                        File.read(local_config_filepath)
+                      end
+
+        YAML.safe_load(raw_configs || '{}') || {}
+      end
+
+      def app_secrets_s3
+        @app_secrets_s3 ||= Identity::Hostdata.app_secrets_s3(logger: @logger, s3_client: @s3_client)
+      end
+
+      def app_configuration_s3_path
+        "/%<env>s/#{app_configuration_path_component}/v1/application.yml"
+      end
+
+      def role_configuration_s3_path
+        return if role_configuration_filename.nil?
+        "/%<env>s/#{app_configuration_path_component}/v1/#{role_configuration_filename}"
+      end
+
+      def app_configuration_path_component
+        return 'idp' if Identity::Hostdata.instance_role == 'worker'
+        Identity::Hostdata.instance_role
+      end
+
+      def role_configuration_filename
+        case Identity::Hostdata.instance_role
+        when 'idp'
+          'web.yml'
+        when 'worker'
+          'worker.yml'
+        end
+      end
+
+      def deep_merge(hash_a, hash_b)
+        hash_a.merge(hash_b) do |key, a_val, b_val|
```

Comment:
Would [Hash#deep_merge](https://api.rubyonrails.org/classes/Hash.html#method-i-deep_merge) work here?

Reply:
It would. That is an activesupport extension. If we want to pull in activesupport we can use that here and we can use `Hash#except` above to get rid of the environment keys.

Reply:
So I guess the question is is there a reason ActiveSupport is not a dep currently? It will end up included anyway if this gem is consumed by a Rails app.

Reply:
Ah, I got confused about which repo I'm in.
I'd be fine with `'activesupport', '~> 6.1'` or whatever, yeah

Reply:
+1, realistically this gem is inside of rails most of the time, seems fine to include it and require the specific extensiion file (not activesupport/all)

Reply:
Active support added to the gemspec!